lib/usb-device.js

```diff
@@ -210,72 +211,68 @@ class USBDevice {
     return fileData;
   }
 
-  // some fun stuff necessary to format buffers for pushing files
-  _constructFileBuffer (fileData, fileStats, devicePath) {
-    let currentPosition = 0; // current position in the file data buffer
-    let remaining = fileData.length;
-    let maxSize = 65536; // we can't send a DATA with more than 64k
+  // loops through a file and sends it to the device with all the required
+  // adb data and formatting in place, called by _sendFile
+  async _fileLoop(filePath, fileStats, devicePath, localId, remoteId) {
     let sendBuffer = new Buffer(`${devicePath},${fileStats.mode}`);
-    let sizeBuffer = new Buffer(4);
-    while (remaining > maxSize) {
-      sizeBuffer.writeUInt32LE(maxSize, 0);
-      let chunk = new Buffer(maxSize);
-      fileData.copy(chunk, 0, currentPosition, currentPosition + maxSize);
-      sendBuffer = Buffer.concat([sendBuffer, dataIndicator, sizeBuffer, chunk]);
-      remaining -= maxSize;
-      currentPosition += maxSize;
-    }
-    sizeBuffer.writeUInt32LE(remaining, 0);
-    let chunk = new Buffer(remaining);
-    fileData.copy(chunk, 0, currentPosition, currentPosition + remaining);
-    sendBuffer = Buffer.concat([sendBuffer, dataIndicator, sizeBuffer, chunk]);
-    let doneBuffer = new Buffer("DONE");
-    let mTimeBuffer = new Buffer(4);
-    mTimeBuffer.writeUInt32LE(fileStats.mtime.getTime() / 1000, 0);
-    sendBuffer = Buffer.concat([sendBuffer, doneBuffer, mTimeBuffer]);
-    return sendBuffer;
-  }
-
-  async _fileLoop (dataBuffer, localId, remoteId) {
-    let remaining = dataBuffer.length;
-    let amountSent = 0;
-    let payloadBuffer = null;
-    while (remaining !== 0) {
-      if (remaining > maxRead) {
-        payloadBuffer = new Buffer(maxRead);
-        dataBuffer.copy(payloadBuffer, 0, amountSent, amountSent + maxRead);
+    let fd = await openAsync(filePath, 'r');
+    let remaining = fileStats.size;
+    let currentPosition = 0;
+    let appendDataSize = true;
+    // DATA<size> can only be 64k or less
+    let dataAmount = remaining > MAX_SIZE ? MAX_SIZE : remaining;
+    while (remaining > 0) {
+      if (appendDataSize) {
+        let sizeBuffer = new Buffer(4);
+        sizeBuffer.writeUInt32LE(dataAmount);
+        sendBuffer = Buffer.concat([sendBuffer, dataIndicator, sizeBuffer]);
+        appendDataSize = false;
+      }
+      let amountToRead = null;
+      if (sendBuffer.length === 0) {
+        amountToRead = dataAmount > MAX_READ ? MAX_READ : dataAmount;
       } else {
-        payloadBuffer = new Buffer(remaining);
-        dataBuffer.copy(payloadBuffer, 0 , amountSent, dataBuffer.length);
+        amountToRead = MAX_READ - sendBuffer.length;
```

Comment:
Nope.  This is the case where we're at the end of one `DATA<size>` and the start of another, so our packet has `fileDataDATA<size>moreFileData` where `moreFileData` is an amount less than `MAX_READ` since there is some data at the front of the packet before that `DATA<size>`.